ddtrace/contrib/pytest/plugin.py

```diff
@@ -0,0 +1,117 @@
+import pytest
+
+from ddtrace import config as ddconfig
+
+from ...ext import SpanTypes, test
+from ...ext.provider import Provider
+from ...pin import Pin
+from .constants import FRAMEWORK, HELP_MSG
+
+
+def is_enabled(config):
+    """Check if the ddtrace plugin is enabled."""
+    return config.getoption("ddtrace") or config.getini("ddtrace")
+
+
+def _extract_span(item):
+    """Extract span from `pytest.Item` instance."""
+    return getattr(item, "_datadog_span", None)
+
+
+def _store_span(item, span):
+    """Store span at `pytest.Item` instance."""
+    setattr(item, "_datadog_span", span)
+
+
+def pytest_addoption(parser):
+    """Add ddtrace options."""
+    group = parser.getgroup("ddtrace")
+
+    group._addoption(
+        "--ddtrace",
+        action="store_true",
+        dest="ddtrace",
+        default=False,
+        help=HELP_MSG,
+    )
+
+    parser.addini("ddtrace", HELP_MSG, type="bool")
+
+
+def pytest_configure(config):
+    config.addinivalue_line("markers", "dd_tags(**kwargs): add tags to current span")
+
+    if is_enabled(config):
+        Pin(tags=Provider.from_env().astags(), _config=ddconfig.pytest).onto(config)
+
+
+def pytest_sessionfinish(session, exitstatus):
+    """Flush open tracer."""
+    pin = Pin.get_from(session.config)
+    if pin is not None:
+        pin.tracer.shutdown()
+
+
+@pytest.fixture(scope="function")
+def ddspan(request):
+    pin = Pin.get_from(request.config)
+    if pin:
+        return _extract_span(request.node)
+
+
+@pytest.hookimpl(tryfirst=True, hookwrapper=True)
+def pytest_runtest_protocol(item, nextitem):
+    pin = Pin.get_from(item.config)
+    if pin is None:
+        yield
+        return
+
+    with pin.tracer.trace(SpanTypes.TEST.value, resource=item.nodeid, span_type=SpanTypes.TEST.value) as span:
```

Comment:
IMO users should use the same service name for which they are running the test suite.

Reply:
Yeah I'm thinking of a fallback in case one isn't defined by the user

Reply:
How should it work - via `ddtrace.config["pytest"]["service_name"]`?

Reply:
it'd be like

```
from ..trace_utils import int_service
config._add(_default_service="pytest")

....
pin.tracer.trace(service=int_service(pin, config.pytest))
```

This will use `"pytest"` as the default if any of `config.pytest.service`/`DD_PYTEST_SERVICE` or `DD_SERVICE`/`config.service` are not defined.